File: src/simulation/kernel/SnapshotRegistry.ts
export type SnapshotProvider = () => unknown;

function ordinalCompare(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

function isolate(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(isolate);
  if (value && typeof value === 'object') {
    const source = value as Record<string, unknown>;
    const output: Record<string, unknown> = {};
    for (const key of Object.keys(source)) output[key] = isolate(source[key]);
    return output;
  }
  return value;
}

export class SnapshotRegistry {
  private readonly providers = new Map<string, SnapshotProvider>();

  register(id: string, provider: SnapshotProvider): void {
    if (id.trim().length === 0) throw new Error('snapshot provider id must not be empty');
    if (this.providers.has(id)) throw new Error(`duplicate snapshot provider: ${id}`);
    this.providers.set(id, provider);
  }

  capture(id: string): unknown {
    const provider = this.providers.get(id);
    if (!provider) throw new Error(`unknown snapshot provider: ${id}`);
    return isolate(provider());
  }

  captureAll(): Readonly<Record<string, unknown>> {
    const output: Record<string, unknown> = {};
    for (const id of this.listIds()) output[id] = this.capture(id);
    return Object.freeze(output);
  }

  listIds(): readonly string[] {
    return Object.freeze([...this.providers.keys()].sort(ordinalCompare));
  }
}

File: cpp/engine/src/core/SnapshotRegistry.cpp
#include <civic/core/SnapshotRegistry.hpp>

#include <utility>

namespace civic {

Result<void> SnapshotRegistry::registerProvider(std::string id, SnapshotProvider provider) {
    if (!utf16_detail::validUtf8AndHasNonEcmaTrimCodePoint(id)) {
        return std::unexpected(make_error(ErrorCode::invalid_argument, "snapshot provider id must not be empty"));
    }
    if (providers_.contains(id)) {
        return std::unexpected(make_error(ErrorCode::invalid_argument, "duplicate snapshot provider: " + id));
    }
    providers_.emplace(std::move(id), std::move(provider));
    return {};
}

Result<std::string> SnapshotRegistry::capture(std::string_view id) const {
    const auto provider = providers_.find(id);
    if (provider == providers_.end()) {
        return std::unexpected(make_error(ErrorCode::invalid_argument, "unknown snapshot provider: " + std::string{id}));
    }
    return provider->second();
}

Result<std::map<std::string, std::string, Utf16OrdinalLess>> SnapshotRegistry::captureAll() const {
    std::map<std::string, std::string, Utf16OrdinalLess> output;
    // Freeze the provider set before invoking providers so capture-all matches the TypeScript registry semantics.
    const auto ids = listIds();
    for (const auto& id : ids) {
        auto captured = capture(id);
        if (!captured) return std::unexpected(captured.error());
        output.emplace(id, std::move(*captured));
    }
    return output;
}

std::vector<std::string> SnapshotRegistry::listIds() const {
    std::vector<std::string> ids;
    ids.reserve(providers_.size());
    for (const auto& [id, provider] : providers_) {
        (void)provider;
        ids.push_back(id);
    }
    return ids;
}

} // namespace civic
